docs(README): provide detail on setup, usage, and implementation (#161)
ipfs/go-bitswap: README.md

## Table of Contents

- [Background](#background)
- [Install](#install)
- [Usage](#usage)
- [Implementation](#implementation)
- [Contribute](#contribute)
- [License](#license)

## Background

> 

Reviewer: > Bitswap is a message based protocol, as opposed to response-reply

Not part of this PR but should this be `Bitswap is a message based protocol, as opposed to request-response`?

	"github.com/libp2p/go-libp2p-core/host"
)

var ctx context.Context
var host host.Host
var router routing.ContentRouting
var bstore blockstore.Blockstore

network := bsnet.NewFromIPFSHost(host, router)
exchange := bitswap.New(ctx, network, bstore)
```

Parameter Notes:

1. `ctx` is just the parent context for all of Bitswap
2. `network` is a network abstraction provided to Bitswap on top
of libp2p & content routing. 
3. `bstore` is an IPFS blockstore

### Get A Block Synchronously

```golang
var c cid.Cid
var ctx context.Context
var exchange bitswap.Bitswap

block, err := exchange.GetBlock(ctx, c)
```

Parameter Notes:

1. `ctx` is the context for this request, which can be cancelled to cancel the request
2. `c` is the content ID of the block you're requesting

### Get Several Blocks Asynchronously

```golang
var cids []cid.Cid
var ctx context.Context
var exchange bitswap.Bitswap

blockChannel, err := exchange.GetBlocks(ctx, cids)
```

Parameter Notes:

1. `ctx` is the context for this request, which can be cancelled to cancel the request
2. `cids` is an slice of content IDs for the blocks you're requesting

### Get Related Blocks Faster With Sessions

> In IPFS, content blocks are often connected to each other through a MerkleDAG. If you know ahead of time that block requests are related, Bitswap can make several optimizations internally in how it requests those blocks in order to get them faster. To facilitate this, Bitswap provides a mechanism called a Bitswap session to designate a series of block requests being part of a single higher level operation. You should initialize a bitswap session any time you intend to make a series of block requests that related -- and whose responses are likely to come from the same peers. 

Reviewer: ```suggestion
In IPFS, content blocks are often connected to each other through a MerkleDAG. If you know ahead of time that block requests are related, Bitswap can make several optimizations internally in how it requests those blocks in order to get them faster. To facilitate this, Bitswap provides a mechanism called a Bitswap session to designate a series of block requests being part of a single higher level operation. You should initialize a bitswap session any time you intend to make a series of block requests that are related -- and whose responses are likely to come from the same peers. 
```

```golang
var ctx context.Context
var cids []cids.cid
var exchange bitswap.Bitswap

session := exchange.NewSession(ctx)
blocksChannel, err := session.GetBlocks(ctx, cids)
// later
var relatedCids []cids.cid
relatedBlocksChannel, err := session.GetBlocks(ctx, relatedCids)
```

> Note that new session returns an interface with a GetBlock and GetBlocks method that have the same signature as the overall Bitswap exchange.

Reviewer: ```suggestion
Note that `NewSession` returns an interface with a `GetBlock` and `GetBlocks` method that have the same signature as the overall Bitswap exchange.
```

### Tell bitswap a new block was added to the local datastore

```golang
var blk blocks.Block
var exchange bitswap.Bitswap

err := exchange.HasBlock(blk)
```

## Implementation

The following diagram outlines the major tasks Bitswap handles, and their consituent components:

![Bitswap Components](./docs/go-bitswap.png)

### Sending Blocks

> Internally, when a message with a wantlist is received, it is sent to the

Reviewer: Again not part of this PR, but I'd suggest changing this paragraph to avoid using the passive voice

to be sent, and send it off. The number of task workers is limited by a constant
factor.

### Requesting Blocks

> Client requests for new blocks are handled by the want manager, for every new

Reviewer: Same here, avoid using passive voice

messages. The same process occurs when the client receives a block and sends a
cancel message for it.

### Sessions

> Sessions track related requests for blocks, and attempt to optimize transfer speed and reduce the number of duplicate blocks sent across the network. The basic optimization of sessions is to limit asks for blocks to the peers most likely to have that block and most likely to respond quickly. This is accomplished by tracking who responds to each block request, and how quickly they respond, and then taking optimizing future requests with that data. Ideally, you want to ask peers who've already responded, and responded quickly, and distribute requests among them to prevent duplicates.

Reviewer: ```suggestion
Sessions track related requests for blocks, and attempt to optimize transfer speed and reduce the number of duplicate blocks sent across the network. The basic optimization of sessions is to limit asks for blocks to the peers most likely to have that block and most likely to respond quickly. This is accomplished by tracking who responds to each block request, and how quickly they respond, and using that information to optimize future requests. Ideally, you want to ask peers who've already responded, and responded quickly, and distribute requests among them to avoid receiving many duplicate blocks.
```

### Finding Providers

When bitswap can't find a connected peer who already has the block it wants, it falls back to querying a content routing system (a DHT in IPFS's case) to try to locate a peer with the block.

> Bitswap routes these requests through the ProviderQueryManager system, which rate limits these requests and also deduplicates in process requests.

Reviewer: ```suggestion
Bitswap routes these requests through the ProviderQueryManager system, which rate limits these requests and also deduplicates in-process requests.
```

### Providing

> As a bitswap client receives blocks, by default it announces them on the provided content routing system (again, a DHT in most cases). This system can be turned off by passing `bitswap.ProvideEnabled(false)` as an extra parameter when initializing Bitswap. IPFS currently has it's own experimental provider system ([go-ipfs-provider](https://github.com/ipfs/go-ipfs-provider)) which will eventually replace bitswap's system entirely.

Reviewer: ```suggestion
As a bitswap client receives blocks, by default it announces them on the provided content routing system (again, a DHT in most cases). This system can be turned off by passing `bitswap.ProvideEnabled(false)` as an extra parameter when initializing Bitswap. IPFS currently has its own experimental provider system ([go-ipfs-provider](https://github.com/ipfs/go-ipfs-provider)) which will eventually replace bitswap's provider system entirely.
```
